content/post/2021-10-07-input-checking/index.Rmd

```diff
@@ -0,0 +1,172 @@
+# knitr hook to make images output use Hugo options
+knitr::knit_hooks$set(
+  plot = function(x, options) {
+    hugoopts <- options$hugoopts
+    paste0(
+      "{{<figure src=",
+      '"', x, '" ',
+      if (!is.null(hugoopts)) {
+        glue::glue_collapse(
+          glue::glue('{names(hugoopts)}="{hugoopts}"'),
+          sep = " "
+        )
+      },
+      ">}}\n"
+    )
+  }
+)
+
+```
+
+## Introduction: the dangers of not checking function inputs
+
+R functions and R packages are a convenient way to share code with the rest of the world
+but it is generally not possible to know how, or with what precise aim in mind, others will
+use your code. For example, they might try to use it on objects that your function was not
+designed for. Let's imagine we have written a short function to compute the geometric mean:
+
+```{r}
+geometric_mean <- function(...) {
+  
+  return(prod(...)^(1/...length()))
+  
+}
+```
+
+When you tested the function yourself, anything seemed fine:
+
+```{r}
+geometric_mean(2, 8)
+
+geometric_mean(4, 1, 1/32)
+```
+
+But a different person using your function might expose it to situations it was not prepared to handle, resulting in cryptic errors or undefined behaviour:
+
+```{r, error = TRUE}
+# Input with factors instead of numerics
+geometric_mean(factor(2), 8)
+
+# Input with negative values
+geometric_mean(-1, 5)
+
+# Input with NAs
+geometric_mean(2, 8, NA)
+```
+
+Or worse, it could give an incorrect output:
+
+```{r}
+geometric_mean(c(2, 8))
+```
+
+Because of this, you need to make sure you return clear errors whenever your functions receives input it was not designed for.
+In this blog post, we review a range of approaches to help you check your function inputs and discuss some potential future developments.
+
+## Checking function inputs using base R
+
+There is a built-in mechanism to check input values in base R: `stopifnot()`.
+You can see it [used](https://github.com/wch/r-source/blob/79298c499218846d14500255efd622b5021c10ec/src/library/stats/R/approx.R#L78) [throughout](https://github.com/wch/r-source/blob/79298c499218846d14500255efd622b5021c10ec/src/library/stats/R/cor.R#L36) [R](https://github.com/wch/r-source/blob/79298c499218846d14500255efd622b5021c10ec/src/library/graphics/R/smoothScatter.R#L47) [source](https://github.com/wch/r-source/blob/79298c499218846d14500255efd622b5021c10ec/src/library/base/R/srcfile.R#L23) [code](https://github.com/wch/r-source/blob/79298c499218846d14500255efd622b5021c10ec/src/library/base/R/parse.R#L65).
+As its name suggests, it will *stop* the function execution *if* an object does *not* pass some tests.
+
+```{r, error = TRUE}
+say_hello <- function(name) {
+  stopifnot(is.character(name))
+  paste("Hello", name)
+}
+
+say_hello("Bob")
+say_hello(404)
+```
+
+However, as you can see in this example, the error message is not in plain English but contains some code instructions.
+This can hinder understanding of the issue.
```

Comment:
Should https://style.tidyverse.org/error-messages.html be linked somewhere?

Reply:
It is linked just below, at line 109, as a footnote :wink:. Do you think it would be better here?

Reply:
Then maybe more prominent in the text? :sweat_smile: :woman_facepalming:

Reply:
You mean I should promote the footnote to an actual sentence in the text?

Reply:
yes, but you are free to disagree!

Reply:
I do think it fits better as a footnote. This is not exactly part of the main topic of the post but more like an aside.